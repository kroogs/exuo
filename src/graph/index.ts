/*
 * Copyright Â© 2020 Ty Dira <ty@dira.dev>

 * This file is part of Exuo.

 * Exuo is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.

 * Exuo is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.

 * You should have received a copy of the GNU Affero General Public License
 * along with Exuo.  If not, see <https://www.gnu.org/licenses/>.
 */

export * from './useGraph'

export * from './models/factories'
export * from './models/Graph'
export * from './models/Node'

export * from './views/EdgeList'
export * from './views/LabelEditor'
export * from './views/NodeActions'
export * from './views/NodeLayout'
export * from './views/NodeListItem'
export * from './views/NodeViewer'
