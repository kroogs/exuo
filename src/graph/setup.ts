/*
 * Copyright Â© 2020 Ty Dira <ty@dira.dev>
 *
 * This file is part of Exuo.

 * Exuo is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.

 * Exuo is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.

 * You should have received a copy of the GNU Affero General Public License
 * along with Exuo.  If not, see <https://www.gnu.org/licenses/>.
 */

import React from 'react'
import { Instance } from 'mobx-state-tree'
import { useObserver } from 'mobx-react-lite'

import { Graph } from './models'

export const graphContext = React.createContext<Instance<typeof Graph>>(null)
export const GraphProvider: React.FunctionComponent<{
  value?: Instance<typeof Graph>
}> = ({ children, value }) =>
  React.createElement(
    graphContext.Provider,
    { value: value ?? Graph.create() },
    children,
  )

export function useGraph<S>(selector: (s: Instance<typeof Graph>) => S): S {
  const graph = React.useContext(graphContext)

  if (graph) {
    return useObserver(() => selector(graph))
  }

  throw Error('Cannot use graph before setup')
}
