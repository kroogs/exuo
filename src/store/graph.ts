/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 * Copyright Â© 2020 Justin Krueger */

import { v4 as uuid } from 'uuid'
import {
  types as t,
  SnapshotIn,
  SnapshotOrInstance,
  Instance,
  IAnyType,
  IAnyModelType,
  IAnyStateTreeNode,
} from 'mobx-state-tree'

// TODO
// improve TS types
// types from configuration
// custom default graph type
// getEdgeMapTypes()

type EdgeTypeFn = () => IAnyType

export const edgeMapFactory = (getEdgeType: EdgeTypeFn): IAnyModelType =>
  t
    .model('EdgeMap', {
      edgeMap: t.map(t.array(t.reference(t.late(getEdgeType)))),
    })
    .actions(self => ({
      hasEdge(tag: string, target: Instance<IAnyModelType>): boolean {
        return Boolean(self.edgeMap.get(tag)?.includes(target))
      },
    }))
    .actions(self => ({
      addEdge(tag: string, target: Instance<IAnyModelType>) {
        if (!self.edgeMap.has(tag)) {
          self.edgeMap.set(tag, [])
        }

        self.edgeMap.get(tag)?.push(target)
      },

      getEdgeTag(tag: string): Array<Instance<IAnyModelType>> {
        return self.edgeMap.get(tag) ?? []
      },

      removeEdge(tag: string, target: Instance<IAnyModelType>) {
        if (!self.hasEdge(tag, target)) {
          throw Error(
            `No edge target '${target}' for tag '${tag}' on '${self}'`,
          )
        }

        self.edgeMap.get(tag)?.remove(target)
      },
    }))

export const nodeFactory = (
  getEdgeType: EdgeTypeFn,
  composeModels: Array<IAnyModelType | EdgeTypeFn> = [],
): IAnyModelType =>
  t.compose(
    // @ts-ignore
    ...[
      edgeMapFactory,
      ...composeModels,
      t.model({
        id: t.identifier,
      }),
    ].map(item => (typeof item === 'function' ? item(getEdgeType) : item)),
  )

const Node = nodeFactory(() => Node)

// const NodeType = t.model('NodeType', {
//   name: t.identifier,
//   compose: t.maybe(t.array(t.late((): IAnyType => t.reference(NodeType)))),
//   props: t.array(t.array(t.string)),
// })

export function graphFactory(
  initialNodeModels: Record<string, IAnyModelType> = { Node },
): IAnyModelType {
  const modelCache = {
    ...initialNodeModels,
  }

  const modelValues = Object.values(modelCache)

  return modelCache.Node.props({
    nodesById: t.map(
      modelValues.length > 1 ? t.union(...modelValues) : modelCache.Node,
    ),
    // typesById: t.map(NodeType),
  })
    .named('Graph')
    .actions(self => {
      // const createNodeTypeModel = (typeName: string): void => {
      //   const typeModel = self.typesById.get(typeName)
      //   if (!typeModel) {
      //     throw Error(`No type named '${typeName}'`)
      //   }

      //   const typeModelProps = Object.fromEntries(typeModel.props)
      //   typeModelCache[typeModel.name] = t
      //     .compose(Node, t.model(typeModelProps))
      //     .named(typeModel.name)
      // }

      return {
        createNode(
          modelName: string,
          props: Omit<SnapshotIn<IAnyModelType>, 'id'>,
        ): Instance<IAnyModelType> {
          if (!modelCache[modelName]) {
            throw Error(`No model named '${modelName}'`)
          }

          const instance = modelCache[modelName].create({
            ...props,
            id: uuid(),
          })

          self.nodesById.put(instance)

          return instance
        },
      }
    })
}

export const Graph = graphFactory()
