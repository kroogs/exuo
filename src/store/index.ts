/*
 * Copyright Â© 2020 Ty Dira <ty@dira.dev>

 * This file is part of Exuo.

 * Exuo is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.

 * Exuo is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.

 * You should have received a copy of the GNU Affero General Public License
 * along with Exuo.  If not, see <https://www.gnu.org/licenses/>.
 */

export * from './StoreProvider'
export * from './useStore'
export * from './persist'

export * from './models/Store'

export * from './components/InstanceViewer'
export * from './components/PropertyEditor'
