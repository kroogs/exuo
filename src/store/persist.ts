/*
 * Copyright Â© 2020 Ty Dira <ty@dira.dev>

 * This file is part of Exuo.

 * Exuo is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.

 * Exuo is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.

 * You should have received a copy of the GNU Affero General Public License
 * along with Exuo.  If not, see <https://www.gnu.org/licenses/>.
 */

import {
  Instance,
  onPatch,
  getSnapshot,
  getMembers,
  applySnapshot,
  IAnyModelType,
  isStateTreeNode,
} from 'mobx-state-tree'
import Dexie from 'dexie'

export async function persist(
  instance: Instance<IAnyModelType>,
): Promise<void> {
  const db = new Dexie('default')
  const tableNames = Object.keys(getMembers(instance).properties)
  const tableConfig = Object.fromEntries(tableNames.map(name => [name, 'id']))

  db.version(1).stores(tableConfig)

  return Promise.all(
    tableNames.map(name =>
      db
        .table(name)
        .toArray()
        .then(rows => [name, Object.fromEntries(rows.map(r => [r.id, r]))]),
    ),
  ).then(resolved => {
    applySnapshot(instance, Object.fromEntries(resolved))
    onPatch(instance, patch => {
      const [typeName, id, propName] = patch.path.split('/').slice(1)

      if (patch.op === 'add') {
        db.table(typeName).put(getSnapshot(instance[typeName].get(id)))
      } else if (patch.op === 'replace' || patch.op === 'remove') {
        if (propName) {
          let propValue = instance[typeName].get(id)[propName]

          if (isStateTreeNode(propValue)) {
            propValue = getSnapshot(propValue)
          }

          db.table(typeName).update(id, {
            [propName]: propValue,
          })
        } else {
          db.table(typeName).delete(id)
        }
      } else {
        throw Error(`Unknown patch op '${patch.op}'`)
      }
    })
  })
}
