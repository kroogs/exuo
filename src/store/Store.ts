/*
 * Copyright Â© 2020 Ty Dira <ty@dira.dev>

 * This file is part of Exuo.

 * Exuo is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.

 * Exuo is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.

 * You should have received a copy of the GNU Affero General Public License
 * along with Exuo.  If not, see <https://www.gnu.org/licenses/>.
 */

import { types } from 'mobx-state-tree'

import { Graph } from 'graph'
import { History, route } from 'route'

export const Store = types
  .model('Store', {
    graph: Graph,
    history: History,
    location: types.maybe(types.string),
  })
  .actions(self => ({
    afterCreate() {
      const href = window.location.href
      const path = window.location.pathname
      const envUrl = process.env.PUBLIC_URL || '/'
      const index = href.indexOf(envUrl)

      let location

      if (index === 0) {
        location = href.replace(envUrl, '')
      } else if (index > 0 && path.indexOf(envUrl) === 0) {
        location = path.replace(envUrl, '')
      } else {
        location = path

        if (process.env.NODE_ENV === 'development') {
          console.warn(
            `Invalid 'homepage' value '${process.env.PUBLIC_URL}'` +
              ` in package.json, trying '${path}'`,
          )
        }
      }

      self.location = location.replace(/\/+/, '/')
    },
  }))
