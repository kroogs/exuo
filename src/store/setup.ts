/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 * Copyright Â© 2020 Justin Krueger */

import React from 'react'
import { types as t, Instance } from 'mobx-state-tree'

import { Graph } from './graph'
import { persist } from './utils'

export const Root = t.model('Root', {
  graph: Graph,
})

export const initStore = (): Instance<typeof Root> => {
  const store = Root.create({ graph: {} })

  // store.graph.init([persist])

  return store
}

export const storeContext = React.createContext<Instance<typeof Root> | null>(
  null,
)

export const StoreProvider: React.FunctionComponent = ({ children }) =>
  React.createElement(storeContext.Provider, { value: initStore() }, children)

export function useStore<S>(selector: (s: Instance<typeof Root>) => S): S {
  const store = React.useContext(storeContext)

  if (!store) {
    throw Error('Cannot use store before setup.')
  }

  return selector(store)
}
