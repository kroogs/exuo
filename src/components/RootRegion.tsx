/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 * Copyright Â© 2020 Justin Krueger */

import React from 'react'
import { useObserver } from 'mobx-react-lite'
import AccountCircle from '@material-ui/icons/AccountCircle'
import AppBar from '@material-ui/core/AppBar'
import Avatar from '@material-ui/core/Avatar'
import Paper from '@material-ui/core/Paper'
import Badge from '@material-ui/core/Badge'
import Button from '@material-ui/core/Button'
import Grid from '@material-ui/core/Grid'
import DeleteIcon from '@material-ui/icons/Delete'
import FolderIcon from '@material-ui/icons/Folder'
import IconButton from '@material-ui/core/IconButton'
import InputBase from '@material-ui/core/InputBase'
import List from '@material-ui/core/List'
import ListItem from '@material-ui/core/ListItem'
import ListItemAvatar from '@material-ui/core/ListItemAvatar'
import ListItemIcon from '@material-ui/core/ListItemIcon'
import ListItemSecondaryAction from '@material-ui/core/ListItemSecondaryAction'
import ListItemText from '@material-ui/core/ListItemText'
import MailIcon from '@material-ui/icons/Mail'
import Menu from '@material-ui/core/Menu'
import MenuIcon from '@material-ui/icons/Menu'
import MenuItem from '@material-ui/core/MenuItem'
import MoreIcon from '@material-ui/icons/MoreVert'
import NotificationsIcon from '@material-ui/icons/Notifications'
import SearchIcon from '@material-ui/icons/Search'
import TextField from '@material-ui/core/TextField'
import Toolbar from '@material-ui/core/Toolbar'
import Typography from '@material-ui/core/Typography'
import { fade, makeStyles, Theme, createStyles } from '@material-ui/core/styles'

import { useStore } from 'store'
import { EdgeList } from './EdgeList'
import { PropertyList } from './PropertyList'

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      backgroundColor: theme.palette.background.default,
      color: theme.palette.text.primary,
    },
    bar: {
      backgroundColor: theme.palette.common.white,
    },
    grid: {
      width: '100%',
    },
    grow: {
      flexGrow: 1,
    },
    menuButton: {
      marginRight: theme.spacing(2),
    },
    title: {
      display: 'none',
      [theme.breakpoints.up('sm')]: {
        display: 'block',
      },
    },
    search: {
      position: 'relative',
      borderRadius: theme.shape.borderRadius,
      backgroundColor: fade(theme.palette.common.white, 0.15),
      '&:hover': {
        backgroundColor: fade(theme.palette.common.white, 0.25),
      },
      marginRight: theme.spacing(2),
      marginLeft: 0,
      width: '100%',
      [theme.breakpoints.up('sm')]: {
        marginLeft: theme.spacing(3),
        width: 'auto',
      },
    },
    searchIcon: {
      width: theme.spacing(7),
      height: '100%',
      position: 'absolute',
      pointerEvents: 'none',
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'center',
    },
    inputRoot: {
      color: 'inherit',
    },
    inputInput: {
      padding: theme.spacing(1, 1, 1, 7),
      transition: theme.transitions.create('width'),
      width: '100%',
      [theme.breakpoints.up('md')]: {
        width: 200,
      },
    },
    sectionDesktop: {
      display: 'none',
      [theme.breakpoints.up('md')]: {
        display: 'flex',
      },
    },
    sectionMobile: {
      display: 'flex',
      [theme.breakpoints.up('md')]: {
        display: 'none',
      },
    },
  }),
)

interface RootRegionProps {
  vertexId?: string
}

export const RootRegion: React.FunctionComponent<RootRegionProps> = ({
  vertexId,
}) => {
  const store = useStore(store => store.graph)

  console.log('RootRegion vertexId', vertexId)

  const classes = useStyles()
  const [anchorEl, setAnchorEl] = React.useState<null | HTMLElement>(null)
  const [
    mobileMoreAnchorEl,
    setMobileMoreAnchorEl,
  ] = React.useState<null | HTMLElement>(null)

  const isMenuOpen = Boolean(anchorEl)
  const isMobileMenuOpen = Boolean(mobileMoreAnchorEl)

  const handleProfileMenuOpen = (
    event: React.MouseEvent<HTMLElement>,
  ): void => {
    setAnchorEl(event.currentTarget)
  }

  const handleMobileMenuClose = (): void => {
    setMobileMoreAnchorEl(null)
  }

  const handleMenuClose = (): void => {
    setAnchorEl(null)
    handleMobileMenuClose()
  }

  const handleMobileMenuOpen = (event: React.MouseEvent<HTMLElement>): void => {
    setMobileMoreAnchorEl(event.currentTarget)
  }

  const menuId = 'primary-search-account-menu'
  const renderMenu = (
    <Menu
      anchorEl={anchorEl}
      anchorOrigin={{ vertical: 'top', horizontal: 'right' }}
      id={menuId}
      keepMounted
      transformOrigin={{ vertical: 'top', horizontal: 'right' }}
      open={isMenuOpen}
      onClose={handleMenuClose}
    >
      <MenuItem onClick={handleMenuClose}>Profile</MenuItem>
      <MenuItem onClick={handleMenuClose}>My account</MenuItem>
    </Menu>
  )

  const mobileMenuId = 'primary-search-account-menu-mobile'
  const renderMobileMenu = (
    <Menu
      anchorEl={mobileMoreAnchorEl}
      anchorOrigin={{ vertical: 'top', horizontal: 'right' }}
      id={mobileMenuId}
      keepMounted
      transformOrigin={{ vertical: 'top', horizontal: 'right' }}
      open={isMobileMenuOpen}
      onClose={handleMobileMenuClose}
    >
      <MenuItem>
        <IconButton aria-label="show 4 new mails" color="inherit">
          <Badge badgeContent={4} color="secondary">
            <MailIcon />
          </Badge>
        </IconButton>
        <p>Messages</p>
      </MenuItem>
      <MenuItem>
        <IconButton aria-label="show 11 new notifications" color="inherit">
          <Badge badgeContent={11} color="secondary">
            <NotificationsIcon />
          </Badge>
        </IconButton>
        <p>Notifications</p>
      </MenuItem>
      <MenuItem onClick={handleProfileMenuOpen}>
        <IconButton
          aria-label="account of current user"
          aria-controls="primary-search-account-menu"
          aria-haspopup="true"
          color="inherit"
        >
          <AccountCircle />
        </IconButton>
        <p>Profile</p>
      </MenuItem>
    </Menu>
  )

  return useObserver(() => {
    const graph = store.byId.get('default')
    /* const rows: Array<React.ReactElement> = [] */

    /* if (graph) { */
    /*   graph.vertexById.forEach(item => */
    /*     rows.push( */
    /*       <ListItem key={item.id}> */
    /*         <ListItemText>{item.label}</ListItemText> */
    /*       </ListItem>, */
    /*     ), */
    /*   ) */
    /* } */
    if (!graph) {
      throw new Error('no graph')
    }

    return (
      <div className={classes.root}>
        <EdgeList model={graph} />
        <PropertyList model={graph} />
      </div>
    )
  })
}

/* export const RootRegion: React.FunctionComponent = props => { */
/*   const store = useStore(store => store.graph) */

/*   return useObserver(() => { */
/*     const graph = store.byId.get('default') */

/*     if (graph) { */
/*       const rows: Array<React.ReactElement> = [] */
/*       const handleSubmit: React.FormEventHandler<HTMLFormElement> = event => { */
/*         event.preventDefault() */
/*       } */

/*       graph.vertexById.forEach(item => */
/*         rows.push( */
/*           <ListItem key={item.id}> */
/*             <ListItemText>{item.label}</ListItemText> */
/*           </ListItem>, */
/*         ), */
/*       ) */

/*       return ( */
/*         <Flex */
/*           sx={{ */
/*             flexDirection: 'column', */
/*             maxHeight: '100vh', */
/*             position: 'relative', */
/*             overflowY: 'auto', */
/*           }} */
/*         > */
/*           <Box */
/*             sx={{ */
/*               position: 'sticky', */
/*               top: '0', */
/*               bg: 'background', */
/*             }} */
/*           > */
/*             <Typography variant="h3">Delicious Fruit</Typography> */
/*             <form onSubmit={handleSubmit} noValidate autoComplete="off"> */
/*               <TextField label="Add Record" /> */
/*             </form> */
/*           </Box> */

/*           {rows.length > 0 && <List>{rows.reverse()}</List>} */
/*         </Flex> */
/*       ) */
/*     } else { */
/*       return <div>shit</div> */
/*     } */
/*   }) */
/* } */

export default RootRegion
