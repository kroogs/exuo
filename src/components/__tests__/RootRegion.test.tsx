/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 * Copyright Â© 2020 Ty Dira */

import React from 'react'
import { shallow } from 'enzyme'

import RootRegion from '../RootRegion'
import { graphFactory, Node } from 'store/graph'

let nextId = 0
const makeId = (): string => String(++nextId)

const Graph = graphFactory({ Node }, makeId)

test('Renders a basic graph', () => {
  const graph = Graph.create({ id: makeId() })
  const first = graph.createNode()
  const last = graph.createNode()

  let prev = null

  for (let i = 0; i < 5; i++) {
    const node = graph.createNode()

    node.addEdge('last', last)
    node.addEdge('first', first)
    first.addEdge('child', node)

    if (prev) {
      prev.addEdge('next', node)
      node.addEdge('prev', prev)
    }

    prev = node
  }

  expect(shallow(<RootRegion graph={graph} />)).toMatchSnapshot()
})
