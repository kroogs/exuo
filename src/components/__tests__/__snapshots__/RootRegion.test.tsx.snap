// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders a basic graph 1`] = `
<Styled(MuiBox)
  className="makeStyles-root-1"
  display="flex"
  flexDirection="row"
  flexWrap="nowrap"
/>
`;
